### PR TITLE
feat: Dynatrace scaler

### LINKED ISSUE #2411: Provide scaler for Dynatrace metrics
### Proposal

Provide k8s autoscaling based on Dynatrace Metrics

### Scaler Source

Pods

### Scaling Mechanics

https://www.dynatrace.com/support/help/dynatrace-api/environment-api/metric-v2/get-all-metrics

### Authentication Source

https://www.dynatrace.com/support/help/dynatrace-api/basics/dynatrace-api-authentication

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by cyrilico**
Add new scaler for interacting with Dynatrace and its [Get Metric Data Points API](https://docs.dynatrace.com/docs/shortlink/api-metrics-v2-get-datapoints)

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Closes #2411 

Relates to https://github.com/kedacore/keda-docs/pull/1360

### CODE REVIEW

**Comment by JorTurFer** [comment]
Please, rebase your branch @cyrilico , the e2e issue was related with some breaking changes in otel-collector which are already fixed in `main` branch (your tests hasn't been executed yet indeed)

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+			args:
+				- --set-host-group={{.KubernetesClusterName}}
+
+	activeGate:
+		capabilities:
+		- routing
+		- dynatrace-api
+		- metrics-ingest
+
+		group: {{.KubernetesClusterName}}
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:91]: It's a mess, but you cannot use tabs on templating files:
```suggestion
	dynakubeSecretTemplate = `apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
data:
  apiToken: {{.DynatraceToken}}
  dataIngestToken: {{.DynatraceToken}}
`

	dynakubeTemplate = `apiVersion: dynatrace.com/v1beta1
kind: DynaKube
metadata:
name: {{.KubernetesClusterName}}
namespace: {{.TestNamespace}}
spec:
  apiUrl: {{.DynatraceHost}}/api
  networkZone: {{.KubernetesClusterName}}
  oneAgent:
    cloudNativeFullStack:
      args:
        - --set-host-group={{.KubernetesClusterName}}
  activeGate:
    capabilities:
    - routing
    - dynatrace-api
    - metrics-ingest
    group: {{.KubernetesClusterName}}
```

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+  selector:
+    matchLabels:
+      app: {{.DeploymentName}}
+  template:
+    metadata:
+	  annotations:
+		data-ingest.dynatrace.com/inject: "true"
+		dynatrace.com/inject: "true"
+		oneagent.dynatrace.com/inject: "true"
+      labels:
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:156]: same here:
```suggestion
    metadata:
      annotations:
        data-ingest.dynatrace.com/inject: "true"
        dynatrace.com/inject: "true"
        oneagent.dynatrace.com/inject: "true"
      labels:
```

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  triggers:
+    - type: dynatrace
+      metadata:
+        host: "{{.DynatraceHost}}"
+        threshold: "2"
+        activationThreshold: "3"
+		metricSelector: builtin:service.requestCount.total:splitBy():fold
+		from: now-2m
+      authenticationRef:
+        name: {{.TriggerAuthName}}
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:212]:
```suggestion
        threshold: "2"
        activationThreshold: "3"
          metricSelector: builtin:service.requestCount.total:splitBy():fold
          from: now-2m
      authenticationRef:
        name: {{.TriggerAuthName}}
```

**Comment by cyrilico** [comment]
new changes, please rerun

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+	})
+
+	// Create kubernetes resources
+	CreateKubernetesResources(t, kc, testNamespace, data, templates)
+
+	installDynatrace(t)
+
+	data, templates = getDynatraceTemplateData()
+	// Create Dynatrace-specific kubernetes resources
+	CreateKubernetesResources(t, kc, testNamespace, data, templates)
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:255]: CreateKubernetesResources recreates the namespace too (deleting the previous resources)
```suggestion
	KubectlApplyMultipleWithTemplate(t, data, templates)
```

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+          stabilizationWindowSeconds: 10
+  triggers:
+    - type: dynatrace
+      metadata:
+        host: {{.DynatraceHost}}
+        threshold: "2"
+        activationThreshold: "3"
+	metricSelector: "builtin:service.requestCount.total:splitBy():fold"
+	from: now-2m
+      authenticationRef:
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:207]: Indentations xD
```suggestion
        threshold: "2"
        activationThreshold: "3"
        metricSelector: "builtin:service.requestCount.total:splitBy():fold"
        from: now-2m
      authenticationRef:
```

**Comment by JorTurFer** [comment]
@cyrilico  I've merged 2 fixes to the e2e to trigger them again (you'll need to pull the changes :) )

File: pkg/scalers/dynatrace_scaler.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	meta := dynatraceMetadata{}
+	var err error
+
+	host, err := GetFromAuthOrMeta(config, "host")
+	if err != nil {
+		return nil, err
+	}
+	meta.host = host
+
+	token, err := GetFromAuthOrMeta(config, "token")
```
Comment (JorTurFer) [inline pkg/scalers/dynatrace_scaler.go:89]: As the token is a secret, we have to get it from metadata or workload environment variables, but we shouldn't read it from metadata

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+metadata:
+  name: {{.TriggerAuthName}}
+  namespace: {{.TestNamespace}}
+spec:
+  secretTargetRef:
+  secretTargetRef:
+  - parameter: token
+    name: {{.SecretName}}
+    key: token
+`
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:109]: The key is the secret key
```suggestion
spec:
  secretTargetRef:
  - parameter: token
    name: {{.SecretName}}
    key: apiToken
`
```

**Comment by JorTurFer** [comment]
It seems that the query isn't correct. Checking the logs I see lines like this:
```bash
2024-04-28T22:27:24Z	ERROR	dynatrace_scaler	error executing Dynatrace query	{"type": "ScaledObject", "namespace": "dynatrace-test-ns", "name": "dynatrace-test-so", "error": "dynatrace response does not contain any results"}
```

File: tests/scalers/dynatrace/dynatrace_test.go (hunk 7 above)
Comment (cyrilico) [reply tests/scalers/dynatrace/dynatrace_test.go]: I had another secret with a different key, didn't realize it was actually overwriting 🤦 I'll keep the one yeah

**Comment by JorTurFer** [comment]
``` bash
2024-04-29T19:24:29.2799042Z     helper.go:444: Waiting for deployment replicas to hit target. Deployment - dynatrace-test-deployment, Current  - 0, Target - ***
2024-04-29T19:24:29.2801324Z     dynatrace_test.go:***68: 
2024-04-29T19:24:29.2803048Z         	Error Trace:	/__w/keda/keda/tests/scalers/dynatrace/dynatrace_test.go:***68
2024-04-29T19:24:29.2809714Z         	            				/__w/keda/keda/tests/scalers/dynatrace/dynatrace_test.go:***5***
2024-04-29T19:24:29.2811538Z         	Error:      	Should be true
2024-04-29T19:24:29.2812961Z         	Test:       	TestDynatraceScaler
2024-04-29T19:24:29.2814935Z         	Messages:   	replica count should be *** after *** minutes
2024-04-29T19:24:29.2816859Z     dynatrace_test.go:***7***: --- testing scale in ---
2024-04-29T19:24:29.2818565Z     helper.go:6***9: Deleting template: lightLoadTemplate
2024-04-29T19:24:29.2820272Z     helper.go:6***9: Deleting template: heavyLoadTemplate
2024-04-29T19:24:29.2827363Z     helper.go:444: Waiting for deployment replicas to hit target. Deployment - dynatrace-test-deployment, Current  - 0, Target - 0
2024-04-29T19:24:29.2830122Z     helper.go:6***9: Deleting template: dynakubeTemplate
2024-04-29T19:24:29.2831645Z     helper.go:6***: 
2024-04-29T19:24:29.2833111Z         	Error Trace:	/__w/keda/keda/tests/helper/helper.go:6***
2024-04-29T19:24:29.2834855Z         	            				/__w/keda/keda/tests/helper/helper.go:64***
2024-04-29T19:24:29.2836531Z         	            				/__w/keda/keda/tests/helper/helper.go:669
2024-04-29T19:24:29.2838378Z         	            				/__w/keda/keda/tests/scalers/dynatrace/dynatrace_test.go:***6
2024-04-29T19:24:29.2844995Z         	            				/usr/local/go/src/testing/testing.go:***69
2024-04-29T19:24:29.2846677Z         	            				/usr/local/go/src/testing/testing.go:***47
2024-04-29T19:24:29.2848336Z         	            				/usr/local/go/src/testing/testing.go:***589
2024-04-29T19:24:29.2849939Z         	Error:      	Received unexpected error:
2024-04-29T19:24:29.2851963Z         	            	error: error when deleting "/tmp/dynakubeTemplate***8546***9594": resource name may not be empty
2024-04-29T19:24:29.2853921Z         	Test:       	TestDynatraceScaler
2024-04-29T19:24:29.2856331Z         	Messages:   	cannot apply file - error: error when deleting "/tmp/dynakubeTemplate***8546***9594": resource name may not be empty
2024-04-29T19:24:29.2863363Z     helper.go:***48: deleting namespace dynatrace-test-ns
2024-04-29T19:24:29.2865297Z     helper.go:***0***: waiting for namespace dynatrace-test-ns deletion
2024-04-29T19:24:29.2867305Z     helper.go:***0***: waiting for namespace dynatrace-test-ns deletion
2024-04-29T19:24:29.2869292Z     helper.go:***0***: waiting for namespace dynatrace-test-ns deletion
2024-04-29T19:24:29.2871273Z     helper.go:***0***: waiting for namespace dynatrace-test-ns deletion
2024-04-29T19:24:29.2873017Z --- FAIL: TestDynatraceScaler (40***.4***s)
2024-04-29T19:24:29.2874375Z FAIL
2024-04-29T19:24:29.2880254Z FAIL	command-line-arguments	40***.509s
```

File: tests/scalers/dynatrace/dynatrace_test.go [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	MaxReplicaCount         string
+}
+
+const (
+	dynakubeTemplate = `apiVersion: dynatrace.com/v1beta1
+kind: DynaKube
+metadata:
+name: {{.KubernetesClusterName}}
+namespace: {{.TestNamespace}}
+spec:
```
Comment (JorTurFer) [inline tests/scalers/dynatrace/dynatrace_test.go:65]:
```suggestion
metadata:
  name: {{.KubernetesClusterName}}
  namespace: {{.TestNamespace}}
spec:
```

**Comment by JorTurFer** [comment]
Same result:
```
2024-04-29T20:05:14Z	ERROR	dynatrace_scaler	error executing Dynatrace query	{"type": "ScaledObject", "namespace": "dynatrace-test-ns", "name": "dynatrace-test-so", "error": "dynatrace response does not contain any metric series"}
```

Is this metric correct? `"builtin:service.requestCount.total:splitBy():fold"`

**Comment by cyrilico** [comment]
@JorTurFer I am certain the query should return a valid result. Is the host variable still using the `.apps` sub domain as we discussed in the issue? If so, I believe that might be an issue, as testing locally against another Dynatrace instance I have access to, using that URL results in an error, thinking we're trying to use another kind of API: `Unsupported authorization scheme 'Api-Token'. Dynatrace platform APIs require the authorization scheme 'Bearer'. Use the OAuth2.0 Client Credentials Flow to obtain a token.`

**Comment by JorTurFer** [comment]
I've been on a business trip this week, but let me check out your branch and check the current code using the current secrets

**Comment by JorTurFer** [comment]
Hello,
I've been testing this locally and I found some things:
- The dynatrace operator fails because the token doesn't have enough permissions for it
- The job generators aren't generating the load because the values are overridden 
image

For the dynatrace operator issue, I've created a new secret 'DYNATRACE_OPERATOR_TOKEN', please use it for the operator and the other one for metrics (including for reading them by KEDA)

After making these changes, I see that dynatrace operator has created more pods:
image

But the metric isn't still available:
image

I think that the issue can be in something related with the e2e test configuration because I can see that KEDA is querying the value and getting an empty response with status 200 (so from KEDA side seems that works)
image

**Comment by cyrilico** [comment]
Appreciate the help @JorTurFer 🙏 What metrics are available in that page? The one I used is supposed to be a [built-in](https://docs.dynatrace.com/docs/observe-and-explore/metrics-classic/built-in-metrics#request-count), i.e., supported OOTB, so I'm curious as to what exactly is available in that dynatrace environment

**Comment by cyrilico** [comment]
Any news @JorTurFer ? 🙏

**Comment by JorTurFer** [comment]
> Any news @JorTurFer ? 🙏

I have to apologize for not answering before, I've been involved on a huge working peak :( 
I'll try to review this ASAP.  I guess that the issue with the metrics was that they needed more time to be consolidated, maybe we can use an approach like the Azure Monitor one, just querying about any available metric as default and updating the ScaledObject to trigger the scaling. e.g: Imagine that you can return an arbitrary number and you use it for scaling, at the end of the day, we don't need to check that the vendor calculates correctly the metric but just the connectivity

**Comment by cyrilico** [comment]
Thanks, I have noticed in the corporate instance I have access to that Dynatrace does take a minute (sometimes literally) to ingest a data point, so I believe that might be an interesting approach. I'll take a look at the Azure Monitor scaler and try to get some inspiration

**Comment by JorTurFer** [comment]
For instance, influx does it too: https://github.com/kedacore/keda/blob/main/tests/scalers/influxdb/influxdb_test.go#L106-L117

It sets the expected value within the query and just updating the ScaledObject you change the received value.

And this is the Azure log analytics one (I said azure monitor worngly): https://github.com/kedacore/keda/blob/main/tests/scalers/azure/azure_log_analytics/azure_log_analytics_test.go#L113-L123

**Comment by cyrilico** [comment]
@JorTurFer couldn't find an exact map to replicate that mechanism in Dynatrace (other than potentially mocking the API calls, which would invalidate the connection testing part), so I've tried a configurable default value filter available in their query language spec; can we have another e2e test run?

**Comment by JorTurFer** [comment]
Hello,
I'm so sorry due to the long silent from my side and I'd like to apologize, my life's been complicated during last months. During this week and next week I'll be focus on this PR (and others) aiming to include them within next week release. 
Thanks for your contribution :)

**Comment by JorTurFer** [comment]
@cyrilico , I've updated the e2e test to [use custom metrics](https://docs.dynatrace.com/docs/dynatrace-api/environment-api/metric-v2/post-ingest-metrics). We have the full control over them and they don't need the agent or any other service, just POST to ingest and KEDA will read them 😄

**Review by JorTurFer — APPROVED** [review]
LGTM! Thanks for the contribution! ❤️ 
PTAL @zroubalik

**Review by zroubalik — COMMENTED** [review]
This is really great, there is one minor thing, do you think you can refactor the scaler metadata parsing, we have introduced a new simplified way, it would be great to include this change in this scaler. You can see details at https://github.com/kedacore/keda/issues/5797

**Comment by cyrilico** [comment]
done @zroubalik 👌

**Review by zroubalik — COMMENTED** [review]
thanks @cyrilico !

**Comment by satrox28** [comment]
Thank you, makes life easier now.

**Pull request merged**